app/scripts/lib/WeakRefObjectMap.ts

```diff
@@ -0,0 +1,74 @@
+type WeakRefCompatibleObject<T> = {
+  [P in keyof T]: T[P] extends object ? WeakRef<T[P]> : never;
+};
+
+export class WeakRefObjectMap {
+  map: Map<string, WeakRefCompatibleObject<Record<string, object>>>;
+
+  constructor() {
+    this.map = new Map();
+  }
+
+  set(key: string, value: Record<string, object>) {
+    const weakReffedValueObj = {} as WeakRefCompatibleObject<
+      Record<string, object>
+    >;
```

Comment:
Nit: Here it would be safer to use a type annotation rather than a type assertion. Type assertions are an _override_, they provide no feedback if the override stops making sense or is no longer required. We're not getting as much value from the type system if we use an assertion. With an annotation. TypeScript will still type check it against the value and let us know if it's invalid.

See these draft TypeScript guidelines for details on this recommendation: https://github.com/MetaMask/contributor-docs/blob/85e06a3097ffdacb7b2fbda9d074955582660181/docs/typescript.md#provide-a-type-annotation--when-instantiating-an-empty-composite-data-type-value


```suggestion
    const weakReffedValueObj: WeakRefCompatibleObject<
      Record<string, object>
    > = {};
```

Reply:
done here: https://github.com/MetaMask/metamask-extension/pull/23525/commits/94c4976528bb0f83c9815acd22be4a45fa4f5646